internal/translation/integrations/conversion.go

```diff
@@ -0,0 +1,106 @@
+package integrations
+
+import (
+	"context"
+	"fmt"
+
+	"go.mongodb.org/atlas-sdk/v20231115008/admin"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/cmp"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/pointer"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1/common"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1/project"
+)
+
+type Integration struct {
+	project.Integration
+}
+
+func (i Integration) Identifier() interface{} {
+	return i.Type
+}
+
+func NewIntegration(spec *project.Integration) (*Integration, error) {
+	if spec == nil {
+		return nil, nil
+	}
+	if err := cmp.Normalize(spec); err != nil {
+		return nil, fmt.Errorf("failed to normalize integration: %w", err)
+	}
+	return &Integration{Integration: *spec}, nil
+}
+
+func NewIntegrations(specs []project.Integration) ([]Integration, error) {
+	list := make([]Integration, 0, len(specs))
+	for _, spec := range specs {
+		integration, err := NewIntegration(&spec)
+		if err != nil {
+			return nil, err
+		}
+		list = append(list, *integration)
+	}
+	return list, nil
+}
+
+func fromAtlas(in *admin.ThirdPartyIntegration) (*Integration, error) {
+	return NewIntegration(
+		&project.Integration{
+			Type:                     in.GetType(),
+			AccountID:                in.GetAccountId(),
+			Region:                   in.GetRegion(),
+			TeamName:                 in.GetTeamName(),
+			ChannelName:              in.GetChannelName(),
+			URL:                      in.GetUrl(),
+			MicrosoftTeamsWebhookURL: in.GetMicrosoftTeamsWebhookUrl(),
+			UserName:                 in.GetUsername(),
+			ServiceDiscovery:         in.GetServiceDiscovery(),
+			Enabled:                  in.GetEnabled(),
+		},
+	)
+}
+
+func toAtlas(in Integration, ctx context.Context, c client.Client, defaultNS string) (result *admin.ThirdPartyIntegration, err error) {
```

Comment:
I'd rather avoid the translation layer making calls to clients here. We are coupling the atlas side with the Kubernetes side like this.

As an alternative I would make use of the local type to also hold the passwords:

```go
type Integration struct {
	project.Integration
        LicenseKey string
        WriteToken string
        ...
}

Such type is populated properly before the translation.
```

Reply:
We could be more explicit with 2 types:

```go
type Integration struct {
	project.Integration
}

type IntegrationWithCredentials struct {
	Integration
        LicenseKey string
        WriteToken string
        ...
}
```

Then the API calls also need to be explicit:

```go
Get(ctx context.Context, projectID string, integrationType string) (*Integration, error)
List(ctx context.Context, projectID string) ([]Integration, error)
Create(ctx context.Context, projectID string, integrationType string, integration IntegrationWithCredentials, client client.Client, ns string) error
Update(ctx context.Context, projectID string, integrationType string, integration IntegrationWithCredentials, client client.Client, ns string) error
```